presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PrefilterForLimitingAggregation.java

```diff
@@ -137,34 +138,32 @@ public PlanNode visitSort(SortNode sortNode, RewriteContext<Void> context)
         @Override
         public PlanNode visitLimit(LimitNode limitNode, RewriteContext<Void> context)
         {
-            PlanNode source = limitNode.getSource();
-            AggregationNode aggregationNode;
+            PlanNode source = rewriteWith(this, limitNode.getSource());
+            AggregationNode aggregationNode = null;
 
             if (source instanceof ProjectNode && ((ProjectNode) source).getSource() instanceof AggregationNode) {
                 aggregationNode = (AggregationNode) ((ProjectNode) source).getSource();
             }
             else if (source instanceof AggregationNode) {
                 aggregationNode = (AggregationNode) source;
             }
-            else {
-                return limitNode;
-            }
 
-            if (!aggregationNode.getGroupingKeys().isEmpty() && isScanFilterProject(aggregationNode.getSource())) {
+            if (aggregationNode != null &&
+                    !aggregationNode.getGroupingKeys().isEmpty() &&
+                    isScanFilterProject(aggregationNode.getSource()) &&
+                    !isAllLowCardinalityGroupByKeys(aggregationNode, aggregationNode.getSource(), metadata, session, limitNode.getCount())) {
```

Comment:
you should be able to pass the StatsProvider into this optimizer and get the stats that way instead of adding a planner util. 
Something like below (this is pseudocode just to show approximately how to use statsProvider, not actually proper code.)
```
private boolean isAllLowCardinalityGroupByKeys(aggregation, source, limitNode.getcount()) {
    PlanNodeStatsEstimate estimate = statsProvider.getStats source())
    if(!estimate.isConfident()) {
       return false:
    }
   for(groupingKey : aggregation.getGroupingKeys()) {
      if(statsEstimate.getVariableStats(groupingKey).getDistinctValuesCount() > count) {
           return false;
      }
   }
  return true
}

Reply:
I can't find how to get statsProvider :(

Reply:
you can pass StatsCalculator in planoptimizers and wrap it in a CachingStatsProvider.  Why isn't this an iterative optimizer rule though? then you'd have it passed through from there.

Reply:
Oh - I need to pass statsCalculator that can help me create a stats provider lol - this middle tier stuff is super annoying

Reply:
Done

Reply:
Also, instead of checking specifically for ScanFilterProject, I wonder if you want to do something similar to https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java#L213, where you look for the first known and confident distinct values that don't pass through an "expanding node".

Reply:
Thisi s a very specific use case and we only want to support scan/filter/project as the rewrite copies the source. We don't want to do it for complex subselects for now. I will just keep it simple.